### ISSUE TITLE
using functions that pass multiple arguments to callbacks without changing promisification

### ISSUE THREAD

**Issue by greaber** [petkaantonov/bluebird#839]
I have just taken a look at Bluebird 3, and I think the change away from packaging multiple callback arguments into arrays is good, both because of the worry about library authors changing their APIs in a way that they thought was backwards compatible but actually breaking Bluebird-promisified applications and because there are many cases where the second argument is rarely useful, so to the application developer using a Bluebird-promisified API adds the cognitive burden of remembering whether to use `.then` or `.spread` on each function (depending on whether that function has no second argument or an unused one).

But changing the promisification API like Bluebird 3 is going to make promisification more complex and confusing for users.  If there are multiple ways to promisify things, cutting-and-pasting code becomes less reliable.  Also, most people just use `promisifyAll`, so `multiArgs` is like using the nuclear option.  Promisification should be as predictable as possible (which is also why projects like https://github.com/lukeb-uk/node-promise-mysql are bad for the community).

Wouldn't it be possible to instead just introduce variants of `.then` and `.spread` that package all the arguments into an array (which then gets spread in the case of `.spread`)?

Probably I am failing to appreciate some crucial point, so sorry in advance if this is the case.

**Comment by bergus** [comment]
> Wouldn't it be possible to instead just introduce variants of  .then  and  .spread  that package all the arguments into an array?

This makes no sense. `then` and `spreads` are methods of the promise that has already been fulfilled with only the first argument. They don't - can't - know about the other arguments.

**Comment by greaber** [comment]
It is very possible I'm suggesting something nonsensical, but I don't understand your explanation yet.  `.then` and `.spread` are typically invoked before the promise is fulfilled, right?  And the same would be true of hypothetical `.thenMulti` and `.spreadMulti` methods.

**Comment by spion** [comment]
It may be possible to make spread work like then on non-arrays. Not sure how big of a breaking change that would be, but I think it isn't (what was previously an error will now be valid)

**Comment by petkaantonov** [comment]
You need to do it at the promisification level, it cannot be done at client methods like .then or .spread.

while spread can work on non-arrays, it would still be ambiguous as to when an array is meant to be an actual array or when it's ferrying multiargs

**Comment by greaber** [comment]
Can you explain slowly what the obstacle is to using client methods?

Probably you understand already what my suggestion was, but just in case, maybe I should explain it more fully.  I am thinking that the promisified function would return a promise that for the most part would behave like a promise of whatever the first callback argument is.  But there would be extra methods on it that you could use to get the other callback arguments.  For instance `.thenMulti` would take a function and pass all the callback arguments as an array to that function when the promise resolves, and `.spreadMulti` would take a function and pass all the callback arguments to that function as separate arguments when the promise resolves.

**Comment by petkaantonov** [comment]
The client methods don't know about any callback, they are methods on the promise. They only know about the promise state.

**Comment by greaber** [comment]
We are having some sort of miscommunication, or maybe I am just really confused, but I am still clueless about what the obstacle is.  It is true that under my suggestion the resolved promise would include extra data that couldn't be accessed through standard methods like `.then` but only through special Bluebird methods like `.thenMulti`.  Is that part of the issue?  By "the callback arguments" I mean the arguments passed by the original unpromisified function to the callback Bluebird (I assume) is providing behind the scenes.  Is that a point of confusion?

**Comment by petkaantonov** [comment]
Yes I suppose a new field could be add to all promises to store the extra args that then could be accessed with special methods.

But each promise would be 8 bytes heavier, and that's for something that is not needed 99.99999999999% of the time.

The filter option should work just fine for this, given how rarely the second argument is even needed:


```js
Promise.promisifyAll(something, {
    filter: function(name) {
        return name === "theMultiArgMethodIwant";
    },
    multiArgs: true
});
// Rest of the methods
Promise.promisifyAll(something);
```

**Comment by greaber** [comment]
Is it really so uncommon?  I haven't been using promises all that long, and these multi-argument callbacks have already come up twice: with node-mysql's `query` method and with node's `execFile` function (although only in the latter case did I actually care about the second argument).

**Comment by greaber** [comment]
Also, is it not possible to just add the new field if extra args are actually passed?  I understand that this kind of thing can be inefficient in v8, but I would think the cost would only be paid in the case where the extra args were added?

**Comment by petkaantonov** [comment]
With optional properties you either pay for the space anyway or the promise object is no longer a single type 

I didn't mean that multiple arguments is rare, just that their usefulness is rare. For the execFile for example, why do you need to access stderr in success case? I wouldn't even use promisification for that as I prefer to fail in  the case where there is output on stderr but somehow the process still exited successfully.

**Comment by greaber** [comment]
OK, I guess you are right that using the second argument turns out to be somewhat rare and there are adequate workarounds like not using promises or maybe even using the new multiArgs feature, in a targeted way.  I am closing the issue.

**Comment by Krisa** [comment]
@petkaantonov or anyone else, is there some best practices for well known modules, such as Needle? Indeed, Needle returns always 2 arguments when promisified: first, it's the response with many uninteresting things and the second argument is the body of the request. With bluebird 2.x, it was possible to do something as elegant as that (notice the `get(1)` for getting the second argument):
```javascript
return needle.postAsync(post_options.url, post_data, post_options).get(1).then(function(myBody){
    ....
});
```

But now, this ends up like that:
```javascript
return needle.postAsync(post_options.url, post_data, post_options).then(function(myBody){
    myBody = myBody[1]; //hmm....
    ....
});
```
or eventually like that:
```javascript
return needle.postAsync(post_options.url, post_data, post_options).spread(function(resp, myBody){
    ....
});
```

Either way, not great...

**Comment by greaber** [comment]
If I understand correctly, the code you are posting won't even work because Bluebird 3 will only give you the first argument to the callback (unless you use multiArgs).  But I believe needle gives you the body on `resp.body` so you don't need the second argument anyway.

**Comment by petkaantonov** [comment]
```js
return needle.postAsync(post_options.url, post_data, post_options).then(function(res){
    console.log(res.body);
});
```

**Comment by Krisa** [comment]
Thanks, probably better than my alternatives, this allows at least to avoid to change the promisification declaration (to multiArgs).

**Comment by tyscorp** [comment]
```js
return needle.postAsync(post_options.url, post_data, post_options).get('body').then(function(myBody){
    console.log(body);
});
```

**Comment by Krisa** [comment]
Amazing @tyscorp , didn't know we could use `get` this way :)
